core/shared/src/main/scala/scalaz/zio/ZIO.scala

```diff
@@ -1,18 +1,1604 @@
+   */
+  final def const[B](b: => B): ZIO[R, E, B] = self.map(_ => b)
+
+  /**
+   * A variant of `flatMap` that ignores the value produced by this effect.
+   */
+  final def *>[R1 <: R, E1 >: E, B](io: => ZIO[R1, E1, B]): ZIO[R1, E1, B] = self.flatMap(_ => io)
+
+  /**
+   * Sequences the specified effect after this effect, but ignores the
+   * value produced by the effect.
+   */
+  final def <*[R1 <: R, E1 >: E, B](io: => ZIO[R1, E1, B]): ZIO[R1, E1, A] = self.flatMap(io.const(_))
+
+  /**
+   * Sequentially zips this effect with the specified effect using the
+   * specified combiner function.
+   */
+  final def zipWith[R1 <: R, E1 >: E, B, C](that: ZIO[R1, E1, B])(f: (A, B) => C): ZIO[R1, E1, C] =
+    self.flatMap(a => that.map(b => f(a, b)))
+
+  /**
+   * Sequentially zips this effect with the specified effect, combining the
+   * results into a tuple.
+   */
+  final def zip[R1 <: R, E1 >: E, B](that: ZIO[R1, E1, B]): ZIO[R1, E1, (A, B)] =
+    self.zipWith(that)((a, b) => (a, b))
+
+  /**
+   * The moral equivalent of `if (p) exp`
+   */
+  final def when[R1 <: R, E1 >: E](b: Boolean)(implicit ev1: ZIO[R, E, A] <:< ZIO[R1, E1, Unit]): ZIO[R1, E1, Unit] =
+    ZIO.when(b)(self)
+
+  /**
+   * The moral equivalent of `if (p) exp` when `p` has side-effects
+   */
+  final def whenM[R1 <: R, E1 >: E](
+    b: ZIO[R1, Nothing, Boolean]
+  )(implicit ev1: ZIO[R, E, A] <:< ZIO[R1, E1, Unit]): ZIO[R1, E1, Unit] =
+    ZIO.whenM(b)(self)
+
+  /**
+   * Repeats this effect forever (until the first error). For more sophisticated
+   * schedules, see the `repeat` method.
+   */
+  final def forever: ZIO[R, E, Nothing] = self *> self.forever
+
+  /**
+   * Repeats this effect with the specified schedule until the schedule
+   * completes, or until the first failure.
+   * Repeats are done in addition to the first execution so that
+   * `io.repeat(Schedule.once)` means "execute io and in case of success repeat `io` once".
+   */
+  final def repeat[R1 <: R, B](schedule: Schedule[R1, A, B]): ZIO[R1 with Clock, E, B] =
+    repeatOrElse[R1, E, B](schedule, (e, _) => ZIO.fail(e))
+
+  /**
+   * Repeats this effect with the specified schedule until the schedule
+   * completes, or until the first failure. In the event of failure the progress
+   * to date, together with the error, will be passed to the specified handler.
+   */
+  final def repeatOrElse[R1 <: R, E2, B](
+    schedule: Schedule[R1, A, B],
+    orElse: (E, Option[B]) => ZIO[R1, E2, B]
+  ): ZIO[R1 with Clock, E2, B] =
+    repeatOrElseEither[R1, B, E2, B](schedule, orElse).map(_.merge)
+
+  /**
+   * Repeats this effect with the specified schedule until the schedule
+   * completes, or until the first failure. In the event of failure the progress
+   * to date, together with the error, will be passed to the specified handler.
+   */
+  final def repeatOrElseEither[R1 <: R, B, E2, C](
+    schedule: Schedule[R1, A, B],
+    orElse: (E, Option[B]) => ZIO[R1 with Clock, E2, C]
```

Comment:
Why does `R1` need the `with Clock` here but not in the `orElse` above?
I am clearly missing something :)

Reply:
in `repeatOrElseEither` we used Schedule which has `Clock` now